src/types.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * These represent the types in our abstract type system.
+ * We will normalize the raw XTP schema into these recursively defined types.
+ */
+
+export type XtpNormalizedKind =
+  'object' | 'enum' | 'map' | 'array' | 'string' |
+  'int32' | 'int64' | 'float' | 'double' |
+  'boolean' | 'date-time' | 'byte' | 'buffer'
+
+// applies type opts to a type on construction
+function cons(t: XtpNormalizedType, opts?: XtpTypeOpts): XtpNormalizedType {
+  // default them to false
+  t.nullable = (opts?.nullable === undefined) ? false : opts.nullable
+  t.required = (opts?.required === undefined) ? false : opts.required
```

Comment:
debating moving required out of the type. Technically it exists in both places. It remains in an array on the Object but we set this boolean here on the property type for convenience.

Reply:
(as you know!) I suspect this should probably get split into `PropertyDescriptor` and `NormalizedType`, where things like `required`, field-level `description`, `name` (or `namePattern`) live on the `PropertyDescriptor` and things like `kind`, `nullable`, type-level `description` and other type info can live on the `NormalizedType`. This gives you a distinction between "field-level" properties when doing code generation and "type-level" properties. On the codegen side you'd end up iterating over a list of `PropertyDescriptor` objects with type information inside.

(This gets at a pretty big impedance mismatch between JSONSchema and the sort of codegen XTP does – JSONSchema doesn't need to name types, but having a name for any sort of rich type is required in most languages we target. This might be a good place to generate an "inferred name" for any type that's otherwise unnamed by the schema as authored?)

A maybe-more complete sketch:


```ts
interface PropertyDescriptor {
  value: XtpNormalizedType,
  description?: string
  codeExamples: string[]
}

interface NamedPropertyDescriptor extends PropertyDescriptor {
  name: string
  required: boolean
}

interface PatternPropertyDescriptor extends PropertyDescriptor {
  pattern: RegExp
}

// pretty much the same as before, sans "required"
interface XtpTypeOpts {
  nullable?: boolean;
}

// ditto!
export interface XtpNormalizedType extends XtpTypeOpts {
  kind: XtpNormalizedKind;
}

// a little different: note the split between `name` and `inferredName` and the list of `properties`
class ObjectType implements XtpNormalizedType {
  kind: XtpNormalizedKind = 'object';
  // if the object appears in a "naming" position (e.g., listed in components/schemas)
  name?: string;

  // Type-level docs and code examples can go here.
  description: string;
  codeExamples: string;

  // if name is not available we infer a name by building up a path from the last named object
  inferredName: string;
  properties: PropertyDescriptor[]
}
```

Reply:
Thanks for the code! that clears it up. I will give integrating this idea a shot. Though I think this points at some friction i'm experiencing now that we have these types. Basically, when do we use the stuff on the nodes and when do we use the stuff on the types? right now there is a lot of overlap. I think after this change percolates we can pull more of this info off the node level and into the property / type level.

> (This gets at a pretty big impedance mismatch between JSONSchema and the sort of codegen XTP does – JSONSchema doesn't need to name types, but having a name for any sort of rich type is required in most languages we target. This might be a good place to generate an "inferred name" for any type that's otherwise unnamed by the schema as authored?)

It is, it's a small one. but i think the end experience is so much better for the price we may need to pay. I don't think we want to generate an inferred name, though it could be an option in the future in order to make transitioning easier. For the time being, we are enforcing a name by sortof forcing them create a schema. And soon schemas will be able to represent any type. Once we allow people to define more inline then we'll want to consider the naming problem.

Reply:
I don't think it's going to come out exactly like you described, but as i'm looking into other languages definitely seeing that required should be extracted from the type. So thanks for the insight!